prov/verbs/src/verbs_mr.c

```diff
@@ -130,37 +154,292 @@ fi_ibv_mr_reg(struct fid *fid, const void *buf, size_t len,
 			fi_eq_write(&md->domain->util_domain.eq->eq_fid,
 				    FI_MR_COMPLETE, &entry, sizeof(entry), 0);
 	}
-	return 0;
-
-err:
-	free(md);
-	return -errno;
+	return FI_SUCCESS;
 }
 
-static int fi_ibv_mr_regv(struct fid *fid, const struct iovec * iov,
-		size_t count, uint64_t access, uint64_t offset, uint64_t requested_key,
-		uint64_t flags, struct fid_mr **mr, void *context)
+static inline
+int fi_ibv_mr_regattr_check_args(struct fid *fid,
+				 const struct fi_mr_attr *attr,
+				 uint64_t flags)
 {
-	if (count > VERBS_MR_IOV_LIMIT) {
+	if (OFI_UNLIKELY(flags))
+		return -FI_EBADFLAGS;
+	if (OFI_UNLIKELY(fid->fclass != FI_CLASS_DOMAIN))
+		return -FI_EINVAL;
+	if (OFI_UNLIKELY(attr->iov_count > VERBS_MR_IOV_LIMIT)) {
 		VERBS_WARN(FI_LOG_FABRIC,
 			   "iov count > %d not supported\n",
 			   VERBS_MR_IOV_LIMIT);
 		return -FI_EINVAL;
 	}
-	return fi_ibv_mr_reg(fid, iov->iov_base, iov->iov_len, access, offset,
-			requested_key, flags, mr, context);
+	return FI_SUCCESS;
+}
+
+static inline int
+fi_ibv_mr_ofi2ibv_access(uint64_t ofi_access, struct fi_ibv_domain *domain)
+{
+	int ibv_access = 0;
+
+	/* Enable local write access by default for FI_EP_RDM which hides local
+	 * registration requirements. This allows to avoid buffering or double
+	 * registration */
+	if (!(domain->info->caps & FI_LOCAL_MR) &&
+	    !(domain->info->domain_attr->mr_mode & FI_MR_LOCAL))
+		ibv_access |= IBV_ACCESS_LOCAL_WRITE;
+
+	/* Local read access to an MR is enabled by default in verbs */
+	if (ofi_access & FI_RECV)
+		ibv_access |= IBV_ACCESS_LOCAL_WRITE;
+
+	/* iWARP spec requires Remote Write access for an MR that is used
+	 * as a data sink for a Remote Read */
+	if (ofi_access & FI_READ) {
+		ibv_access |= IBV_ACCESS_LOCAL_WRITE;
+		if (domain->verbs->device->transport_type == IBV_TRANSPORT_IWARP)
+			ibv_access |= IBV_ACCESS_REMOTE_WRITE;
+	}
+
+	if (ofi_access & FI_WRITE)
+		ibv_access |= IBV_ACCESS_LOCAL_WRITE;
+
+	if (ofi_access & FI_REMOTE_READ)
+		ibv_access |= IBV_ACCESS_REMOTE_READ;
+
+	/* Verbs requires Local Write access too for Remote Write access */
+	if (ofi_access & FI_REMOTE_WRITE)
+		ibv_access |= IBV_ACCESS_LOCAL_WRITE |
+			      IBV_ACCESS_REMOTE_WRITE |
+			      IBV_ACCESS_REMOTE_ATOMIC;
+
+	return ibv_access;
+}
```

Comment:
It seems fi_ibv_mr_ofi2ibv_access function is independent from enabling MR cache. Can you please move it into a separate patch?

Reply:
Sure, we can. Done

Reply:
It seems this hasn't been moved to a separate patch.

Reply:
> It seems this hasn't been moved to a separate patch.

Sorry, it has been missed during force push from auxiliary repo.

Now it should be there.